keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,421 @@
+[Kueue](https://github.com/kubernetes-sigs/kueue) only when all resources are available for
+every ReplicatedJob within JobSet.
+
+That allows us to leverage the existing integration between Kueue and JobSet while using the
+ExecutionPolicy API.
+
+In the future versions we will discuss potential partial admission of JobSet by Kueue. For example,
+when compute resources are available for the first ReplicatedJob the JobSet can be dispatched by
+Kueue.
+
+### Defaulting/Validation
+
+- ExecutionPolicy is immutable.
+- ExecutionPolicyOrder of `AnyOrder` is default setting.
+- StartupPolicy should be equal to `AnyOrder` when ExecutionPolicy is used.
+  - The StartupPolicy API will be deprecated over the next few JobSet releases, since
+    ExecutionPolicy can be used with ready status of ReplicatedJobs.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Unit Tests
+
+- `controllers`: `01/19/2024` - `30.2%`
+
+We will create a new file called execution_policy for the functionality.
+
+#### Integration tests
+
+- Using "ready" ReplicatedJobs status for JobSet with 2+ ReplicatedJobs
+- Using "succeeded" ReplicatedJobs status for JobSet with 2+ ReplicatedJobs
+- Validate that if the first ReplicatedJob fails, the second ReplicatedJob does not execute.
+- Ensure that when ExecutionPolicy and FailurePolicy sets together, the JobSet will restart
+  the job sequence from the beginning in case of failure.
+- Validate the JobSet controller restart in the middle of the ReplicatedJobs execution when
+  ExecutionPolicy is set.
+
+### Graduation Criteria
+
+<!--
+
+Clearly define what it means for the feature to be implemented and
+considered stable.
+
+If the feature you are introducing has high complexity, consider adding graduation
+milestones with these graduation criteria:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Feature gate][feature gate] lifecycle
+- [Deprecation policy][deprecation-policy]
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+-->
+
+## Implementation History
+
+- Draft KEP: September 25th 2024
+
+## Drawbacks
+
+### Manage quota for Job sequence
+
+It is hard to manage quota for sequence of Job within single JobSet. Usually, users don't want to
+lock resources (e.g. GPUs/TPUs) when the first Job is running.
+
+### Support complex DAGs with JobSet
+
+If users want to create complex DAG workflows, they should not use JobSet for such purpose.
+
+## Alternatives
+
+### Add ExecutionPolicyRule parameter into the StartupPolicy API
```

Comment:
I think, both ExecutionPolicy and StartupPolicy makes sense to me for this feature.
Using different API would just make it simpler to deprecate the legacy API.
Any thoughts @kannon92 @tenzen-y @danielvegamyhre ?

Reply:
Initially I thought the name "StartupPolicy" was too different to be applicable to this use case of serial execution (rather than serial startup), but after thinking more about the implementation details and maintenance of these 2 similar code paths, I think extending the StartupPolicy API to support serial execution would be preferable to building and maintaining a new, similar API. 

The name is close enough to be applicable, and I suppose we can always rename it to a more generic "ExecutionPolicy" if/when we graduate the API to beta/v1.